File: app/src/main/java/com/security/passwordmanager/ui/DataRecyclerView.java
//package com.security.passwordmanager.ui;
//
//import android.content.Context;
//import android.view.View;
//
//import androidx.annotation.IdRes;
//import androidx.annotation.NonNull;
//import androidx.appcompat.app.AppCompatActivity;
//import androidx.lifecycle.ViewModelProvider;
//import androidx.recyclerview.widget.LinearLayoutManager;
//import androidx.recyclerview.widget.RecyclerView;
//
//import com.security.passwordmanager.ActionBottom;
//import com.security.passwordmanager.ActionBottomDialogFragment;
//import com.security.passwordmanager.data.BankCard;
//import com.security.passwordmanager.data.Data;
//import com.security.passwordmanager.data.DataType;
//import com.security.passwordmanager.data.DataViewModel;
//import com.security.passwordmanager.data.Website;
//import com.security.passwordmanager.settings.SettingsViewModel;
//
//import java.util.List;
//
//public abstract class DataRecyclerView {
//
//    protected SettingsViewModel support;
//    protected Context context;
//    protected DataViewModel mDataViewModel;
//
//    protected final RecyclerView mRecyclerView;
//    private final LinearLayoutManager mLayoutManager;
//    protected List<Data> mAccountList;
//
//    protected String key;
//    protected boolean editable;
//
//    private OnScroll onScroll, updateViewScroll;
//
//    protected AppCompatActivity mActivity;
//    public int startCount;
//
//    protected ActionBottomDialogFragment bottomDialogFragment;
//
//    public DataRecyclerView(
//            @NonNull View root,
//            @NonNull Context context,
//            @NonNull DataViewModel viewModel,
//            @IdRes int resId,
//            boolean horizontal) {
//
//        this(root, context, viewModel, resId, horizontal, "");
//    }
//
//    public DataRecyclerView(
//            @NonNull AppCompatActivity activity,
//            @IdRes int resId,
//            boolean horizontal,
//            String key) {
//
//        this(
//                activity.getWindow().getDecorView(),
//                activity.getApplicationContext(),
//                new ViewModelProvider(activity).get(DataViewModel.class),
//                resId,
//                horizontal,
//                key
//        );
//
//        this.mActivity = activity;
//    }
//
//
//    //Главный конструктор
//    public DataRecyclerView(
//            @NonNull View root,
//            @NonNull Context context,
//            @NonNull DataViewModel viewModel,
//            @IdRes int resId,
//            boolean horizontal,
//            String key) {
//
//        this.context = context;
//        this.mDataViewModel = viewModel;
//        this.key = key;
//
//        support = new ViewModelProvider(mActivity).get(SettingsViewModel.class);
//
//        mRecyclerView = root.findViewById(resId);
//        if (horizontal)
//            mLayoutManager = new LinearLayoutManager(context, LinearLayoutManager.HORIZONTAL, false);
//        else
//            mLayoutManager = new LinearLayoutManager(context);
//
//        mRecyclerView.setLayoutManager(mLayoutManager);
//
//        mRecyclerView.addOnScrollListener(new RecyclerView.OnScrollListener() {
//            @Override
//            public void onScrollStateChanged(@NonNull RecyclerView recyclerView, int newState) {
//                super.onScrollStateChanged(recyclerView, newState);
//
//                if (newState == RecyclerView.SCROLL_STATE_DRAGGING && onScroll != null)
//                    onScroll.onScrollChanged();
//            }
//
//            @Override
//            public void onScrolled(@NonNull RecyclerView recyclerView, int dx, int dy) {
//                super.onScrolled(recyclerView, dx, dy);
//
//                if (updateViewScroll != null)
//                    updateViewScroll.onScrollChanged();
//
//                if (onScroll != null)
//                    onScroll.onScrollChanged();
//            }
//        });
//
//        editable = true;
//
//        bottomDialogFragment = ActionBottom.Companion.newInstance(mActivity);
//
//        updateAccountList();
//        startCount = mAccountList.size();
//
//        if (mAccountList.size() == 0)
//            if (getType() == DataType.WEBSITE)
//                mAccountList.add(new Website());
//            else
//                mAccountList.add(new BankCard());
//    }
//
//
//    public void setVisibility(int visibility) {
//        mRecyclerView.setVisibility(visibility);
//    }
//
//    public abstract DataType getType();
//
//
//    public void setOnScroll(OnScroll onScroll) {
//        this.onScroll = onScroll;
//    }
//
//    public void updateViewScroll(OnScroll updateViewScroll) {
//        this.updateViewScroll = updateViewScroll;
//    }
//
//    public void setKey(String key) {
//        this.key = key;
//        updateAccountList();
//    }
//
//    public void setAccountList(List<Data> accountList) {
//        this.mAccountList = accountList;
//        updateRecyclerView();
//    }
//
//    public void scrollToPosition(int position) {
//        mLayoutManager.scrollToPosition(position);
//    }
//
//    public void scrollToEnd() {
//        scrollToPosition(mAccountList.size());
//    }
//
//    public View getChildAt(int index) {
//        return mRecyclerView.getChildAt(index);
//    }
//
//    @Deprecated
//    public int getCurrentPosition() {
//        int position = mLayoutManager.findFirstVisibleItemPosition();
//
//        return position == -1 ? 0 : position;
//    }
//
//    public abstract void updateRecyclerView();
//
//
//    public void addData(Data data) {
//        mAccountList.add(data);
//    }
//
//    public void updateData(int position, Data data) {
//        mAccountList.set(position, data);
//    }
//
//    public Data getData(int position) {
//        return mAccountList.get(position);
//    }
//
//    public int getItemCount() {
//        return mAccountList.size();
//    }
//
//    @Deprecated
//    public Data getCurrentData() {
//        return mAccountList.get(getCurrentPosition());
//    }
//
//
//    public void setEditable(boolean editable) {
//        this.editable = editable;
//    }
//
//
//    protected void updateAccountList() {
//        mAccountList = mDataViewModel.getAccountList(key, getType());
//    }
//
//    public interface OnScroll {
//        void onScrollChanged();
//    }
//}

File: app/src/main/java/com/security/passwordmanager/ui/DataRecyclerView.kt
package com.security.passwordmanager.ui

import android.view.View
import androidx.appcompat.app.AppCompatActivity
import androidx.lifecycle.ViewModelProvider
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import com.security.passwordmanager.data.*
import com.security.passwordmanager.settings.SettingsViewModel

abstract class DataRecyclerView(
    protected val activity: AppCompatActivity,
    protected val recyclerView: RecyclerView,
    private val type: DataType,
    key: String = "",
    val editable: Boolean = true
) {

    //todo попробовать имплементить iterable<View>

    var key = key
        set(value) {
            field = value
            updateAccountList()
        }

    private val layoutManager: LinearLayoutManager =
        recyclerView.layoutManager as LinearLayoutManager

    protected val settings = SettingsViewModel.getInstance(activity)
    protected val dataViewModel = ViewModelProvider(activity)[DataViewModel::class.java]

    protected lateinit var accountList : ArrayList<Data>
    val startCount : Int

    abstract fun updateRecyclerView()

    init {
        updateAccountList()
        startCount = accountList.size

        if (isEmpty()) {
            accountList.add(
                if (type == DataType.WEBSITE) Website()
                else BankCard()
            )
        }
    }

    fun copyAccountList() =
        dataViewModel.copyAccountList(accountList)

    protected fun updateAccountList() {
        accountList = (if (key.isEmpty())
            ArrayList()
        else
            dataViewModel.getAccountList(key, type) as ArrayList<Data>)
        updateRecyclerView()
    }

    fun setVisibility(visibility : Int) {
        recyclerView.visibility = visibility
    }

    fun isEmpty() = accountList.isEmpty()

    fun scrollToPosition(position : Int) = recyclerView.post {
        recyclerView.smoothScrollToPosition(position)
    }

    fun scrollToEnd() = scrollToPosition(itemCount - 1)

    operator fun get(index : Int): View = recyclerView.getChildAt(index)

    private val currentPosition : Int get() {
        val position = layoutManager.findFirstVisibleItemPosition()

        return if (position == -1) 0
        else position
    }

    fun addData(data: Data) {
        accountList.add(data)
        updateRecyclerView()
    }

    open fun getData(position: Int) = accountList[position]

    val itemCount get() = accountList.size

    fun getCurrentData() = getData(currentPosition)
}